File: apps/tv/modules/valence-focus-lift/android/src/main/java/app/valence/tv/focuslift/ValenceFocusLiftView.kt
package app.valence.tv.focuslift

import android.content.Context
import android.graphics.Outline
import android.view.View
import android.view.ViewOutlineProvider
import android.view.ViewTreeObserver
import android.view.animation.DecelerateInterpolator
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.views.ExpoView

private const val LIFT_MS = 200L

private const val RAISED_DP = 12f

/**
 * Lifts what it holds as the remote lands on anything inside it, and lets it down as the remote
 * leaves, hearing the move from Android's own focus system rather than waiting on JavaScript.
 *
 * The lift is a scale on the view as drawn, so the frame React Native lays out is never changed
 * beneath it, and it is raised above its neighbours while lifted so a grown card is not drawn under
 * the one beside it. A shadow, where asked for, is cast from the top of what it holds down to the
 * given height, so a card's picture casts it and not the words beneath. A row in a list grows from
 * its left edge rather than its middle, so its words stay in line with the rows around it.
 */
class ValenceFocusLiftView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  var scale = 1.1f
  var isAnchoredLeft = false
  var shadowHeight = 0f
    set(value) {
      field = value
      invalidateOutline()
    }
  var cornerRadius = 0f
    set(value) {
      field = value
      invalidateOutline()
    }

  private var isLifted = false

  private val density = resources.displayMetrics.density

  private val hearFocus = ViewTreeObserver.OnGlobalFocusChangeListener { _, now ->
    val isComing = now != null && holds(now)

    if (isComing != isLifted) {
      isLifted = isComing
      lift(isComing)
    }
  }

  init {
    clipChildren = false
    outlineProvider = object : ViewOutlineProvider() {
      override fun getOutline(view: View, outline: Outline) {
        val height = (shadowHeight * density).toInt().coerceAtMost(view.height)

        outline.setRoundRect(0, 0, view.width, height, cornerRadius * density)
        outline.alpha = if (shadowHeight > 0f) 0.55f else 0f
      }
    }
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    viewTreeObserver.addOnGlobalFocusChangeListener(hearFocus)
  }

  override fun onDetachedFromWindow() {
    viewTreeObserver.removeOnGlobalFocusChangeListener(hearFocus)
    super.onDetachedFromWindow()
  }

  override fun onSizeChanged(width: Int, height: Int, oldWidth: Int, oldHeight: Int) {
    super.onSizeChanged(width, height, oldWidth, oldHeight)
    pivotX = if (isAnchoredLeft) 0f else width / 2f
    pivotY = height / 2f
  }

  /** Whether a view is this one or somewhere inside it. */
  private fun holds(view: View): Boolean {
    var at: View? = view

    while (at != null) {
      if (at === this) {
        return true
      }

      at = at.parent as? View
    }

    return false
  }

  /** Grows what it holds and raises it, or lets it back down. */
  private fun lift(isUp: Boolean) {
    pivotX = if (isAnchoredLeft) 0f else width / 2f
    animate()
      .scaleX(if (isUp) scale else 1f)
      .scaleY(if (isUp) scale else 1f)
      .translationZ(if (isUp) RAISED_DP * density else 0f)
      .setDuration(LIFT_MS)
      .setInterpolator(DecelerateInterpolator())
      .start()
  }
}

File: apps/tv/modules/valence-focus-lift/ios/ValenceFocusLiftView.swift
import ExpoModulesCore
import UIKit

/// Lifts what it holds the moment the television moves the remote onto it, in step with the system's
/// own focus animation, rather than once JavaScript has heard about the move and answered.
///
/// The lift is a transform on what it holds rather than on itself, so its frame, which React Native
/// sets, is never changed beneath it. A shadow, where asked for, is a layer of its own, kept beneath
/// whatever React Native puts in the view however it inserts it, drawn from a fixed outline — the
/// top of what it holds, down to the given height, so a card's picture casts it and not the words
/// beneath — which costs nothing to draw each frame. A row in a list grows from its
/// left edge rather than its middle, so its words stay in line with the rows around it.
final class ValenceFocusLiftView: ExpoView {
  var scale: CGFloat = 1.1
  var isAnchoredLeft = false
  var shadowHeight: CGFloat = 0 {
    didSet { setNeedsLayout() }
  }
  var cornerRadius: CGFloat = 0 {
    didSet { setNeedsLayout() }
  }

  private let shade = CALayer()
  private var isLifted = false

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    shade.shadowColor = UIColor.black.cgColor
    shade.shadowOpacity = 0
    shade.shadowRadius = 28
    shade.shadowOffset = CGSize(width: 0, height: 18)
    shade.zPosition = -1
    layer.insertSublayer(shade, at: 0)
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    CATransaction.begin()
    CATransaction.setDisableActions(true)
    let outline = CGRect(x: 0, y: 0, width: bounds.width, height: min(shadowHeight, bounds.height))
    shade.frame = bounds
    shade.shadowPath = UIBezierPath(roundedRect: outline, cornerRadius: cornerRadius).cgPath
    CATransaction.commit()
  }

  override func didUpdateFocus(
    in context: UIFocusUpdateContext,
    with coordinator: UIFocusAnimationCoordinator
  ) {
    super.didUpdateFocus(in: context, with: coordinator)
    let isComing = context.nextFocusedView?.isDescendant(of: self) ?? false

    guard isComing != isLifted else { return }

    isLifted = isComing

    if isComing {
      coordinator.addCoordinatedFocusingAnimations({ animation in
        self.lift(to: self.scale, over: animation.duration)
      }, completion: nil)
    } else {
      coordinator.addCoordinatedUnfocusingAnimations({ animation in
        self.lift(to: 1, over: animation.duration)
      }, completion: nil)
    }
  }

  /// Moves what it holds to a size, and the shadow in or out with it, over the system's own time.
  private func lift(to size: CGFloat, over duration: TimeInterval) {
    let grown = CATransform3DMakeScale(size, size, 1)
    let target = isAnchoredLeft
      ? CATransform3DConcat(grown, CATransform3DMakeTranslation(bounds.width / 2 * (size - 1), 0, 0))
      : grown
    let grow = CABasicAnimation(keyPath: "sublayerTransform")
    grow.fromValue = layer.presentation()?.sublayerTransform ?? layer.sublayerTransform
    grow.toValue = target
    grow.duration = duration
    grow.timingFunction = CAMediaTimingFunction(name: .easeOut)
    layer.sublayerTransform = target
    layer.add(grow, forKey: "lift")

    guard shadowHeight > 0 else { return }

    let opacity: Float = size > 1 ? 0.55 : 0
    let fade = CABasicAnimation(keyPath: "shadowOpacity")
    fade.fromValue = shade.presentation()?.shadowOpacity ?? shade.shadowOpacity
    fade.toValue = opacity
    fade.duration = duration
    shade.shadowOpacity = opacity
    shade.add(fade, forKey: "shade")
  }
}
